students/EricAdams/session04/trigrams.py

```diff
@@ -1,101 +1,166 @@
+        raw_input_string = f_input.read()
+
+    # Clean up the input
+    input_string = replace_unwanted_punctuation(
+        raw_input_string, unwanted_punc, substitution_chars)
+
+    # get a list of words, instead of chars
+    input_text = "".join(input_string)
+    input_string = input_text.split(" ")
+
+    # Create the trigram from the cleaned up input
+    trigram = create_the_trigram(input_string)
+
+    # Build the text output.text
+    # Start with a random word from the trigram
+
+    output_list = initial_output_list(trigram)
 
-raw_input_list = []
-input_list = []
-initial_trigram_key = tuple()
-output_list = []
-
-replace_punc = ['-',
-                "'",
-                ',',
-                '.',
-                ')',
-                '(',
-                '"',
-                '?',
-                ';',
-                '\n']
-
-# Put the input file into a list,(by char)
-
-
-with open('./sherlock.txt', 'r') as f_input:
-    raw_input_list = list(f_input.read())
-
-# Build up an input list with unwanted punctuation replaced
-for i, x in enumerate(raw_input_list):
-    if x in replace_punc:
-        input_list.append(' ')
-    else:
-        input_list.append(x)
-
-# get a list of words, instead of chars
-input_text = "".join(input_list)
-input_list = input_text.split(" ")
-# print(input_list)
-
-# Create the trigram from the cleaned up input
-trigram = {}
-# we need a pair hence the len(input_list) - 2 instead of len(input_list) - 1
-for i in range(len(input_list) - 2):
-    # must be a tuple in order to be a key in trigram
-    trigram_key = tuple(input_list[i:i + 2])
-    trigram_value = input_list[i + 2]
-    # this is so nifty, I had to use it
-    trigram.setdefault(trigram_key, []).append(trigram_value)
-
-# Build the text output.text
-# Start with a random word from the trigram
-
-random_number = random.randrange(0, len(trigram) - 1)
-for i, j in enumerate(trigram):
-    if i == random_number:
-        initial_trigram_key = j
-        break
-
-output_list = list(j) + trigram[j]
-while(True):
-    # Next keyword pair = last two elements of output_list
-    new_key = tuple(output_list[-2:])
-    try:
-        trigram_value_list = trigram[new_key]
-    except KeyError:
-        # pick a random word in trigram_value_list
-        print('exiting')
-        break
-    random_trigram_value = random.choice(trigram_value_list)
-    new_value = random_trigram_value
-    output_list = output_list + list(new_key) + [new_value]
-    if len(output_list) > 2500:
-        break
-t = " ".join(output_list)
-# print(t)
-with open('trigrams.txt', 'w') as f_output:
-    f_output.write(t)
+    while(True):
+        # Next keyword pair = last two elements of output_list
+        new_key = tuple(output_list[-2:])
+        try:
+            trigram_followers_list = trigram[new_key]
+        except KeyError:
+            print('exiting')
+            break
+        random_trigram_followers = random.choice(trigram_followers_list)
```

Comment:
careful with your variable names -- this is a single word, so confusing to use a plural name.

I would probably call the list simply "followers" - the plural indicates it's more than one -- and then when you pull out a random one, maybe "follower" or "next_word" or something singular.